src/winhello/WindowsHello.cpp

```diff
@@ -0,0 +1,233 @@
+            return false;
+        }
+
+        if (NCryptOpenKey(*hProv, hKey, keyname->pszName, keyname->dwLegacyKeySpec, 0) != ERROR_SUCCESS) {
+            NCryptFreeBuffer(keyname);
+
+            g_winHelloAvailable = WINDOWSHELLO_NOT_AVAILABLE;
+            return false;
+        }
+
+        NCryptFreeBuffer(keyname);
+
+        QString promptMessage =
+            QCoreApplication::translate("DatabaseOpenDialog", "authenticate to access the database");
+        wchar_t* prompt = new wchar_t[promptMessage.length() * 2];
+        promptMessage.toWCharArray(prompt);
+
+        if (NCryptSetProperty(*hKey,
+                              NCRYPT_USE_CONTEXT_PROPERTY,
+                              reinterpret_cast<PBYTE>(prompt),
+                              (promptMessage.length() + 1) * 2,
+                              0)
+            != ERROR_SUCCESS) {
+
+            g_winHelloAvailable = WINDOWSHELLO_NOT_AVAILABLE;
+            return false;
+        }
+
+        QWidget* activeWindow = QApplication::activeWindow();
+        if (activeWindow != nullptr) {
+            WId windowHandle = activeWindow->winId();
+
+            if (NCryptSetProperty(*hKey,
+                                  NCRYPT_WINDOW_HANDLE_PROPERTY,
+                                  reinterpret_cast<PBYTE>(&windowHandle),
+                                  sizeof(windowHandle),
+                                  0)
+                != ERROR_SUCCESS) {
+
+                g_winHelloAvailable = WINDOWSHELLO_NOT_AVAILABLE;
+                return false;
+            }
+        }
+
+        DWORD requirePin = 1;
+
+        if (NCryptSetProperty(
+                *hKey, NCRYPT_PIN_CACHE_IS_GESTURE_REQUIRED_PROPERTY, reinterpret_cast<PBYTE>(&requirePin), 4, 0)
+            != ERROR_SUCCESS) {
+
+            g_winHelloAvailable = WINDOWSHELLO_NOT_AVAILABLE;
+            return false;
+        }
+
+        g_winHelloAvailable = WINDOWSHELLO_AVAILABLE;
+        return true;
+    }
+
+    void freeNCryptHandles(NCRYPT_PROV_HANDLE* hProv, NCRYPT_KEY_HANDLE* hKey)
+    {
+        NCryptFreeObject(*hKey);
+        NCryptFreeObject(*hProv);
+    }
+} // namespace
+
+bool WindowsHello::isAvailable()
+{
+    if (g_winHelloAvailable == WINDOWSHELLO_UNDEFINED) {
+        NCRYPT_PROV_HANDLE hProvider;
+        NCRYPT_KEY_HANDLE hKey;
+
+        getNCryptHandles(&hProvider, &hKey);
+        freeNCryptHandles(&hProvider, &hKey);
+    }
+
+    return g_winHelloAvailable == WINDOWSHELLO_AVAILABLE;
+}
+
+bool WindowsHello::storeKey(const QString& databasePath, const QByteArray& passwordKey)
+{
+    if (databasePath.isEmpty() || passwordKey.isEmpty() || !isAvailable()) {
+        return false;
+    }
+
+    if (g_encryptedMasterKeys.contains(databasePath)) {
+        return true;
+    }
+
+    NCRYPT_PROV_HANDLE hProvider;
+    NCRYPT_KEY_HANDLE hKey;
+
+    if (!getNCryptHandles(&hProvider, &hKey)) {
+        freeNCryptHandles(&hProvider, &hKey);
```

Comment:
Is this safe? I could imagine that if "getting" them doesn't succeed, "freeing" them won't do any good.

Looks like `NCRYPT_HANDLE` is some some sort of opaque pointer. I guess you could at least default-initialize these to `nullptr` and then check whether they have actual values before trying to free either one of them.